lib/streamlit/components/v2/__init__.py

```diff
@@ -11,3 +11,297 @@
+        When a path-like string is provided, the file MUST reside inside the
+        manifest-declared ``asset_dir`` for this component (configured in the
+        package's ``pyproject.toml``). Relative paths are resolved against the
+        caller's file location. Globs are allowed and must resolve to exactly one
+        file within ``asset_dir``. Attempts to reference files outside ``asset_dir``
+        (including traversal) will raise ``StreamlitAPIException``.
+    js : str | None
+        Either inline JavaScript (string) or a string path/glob to a ``.js`` file,
+        validated with the same rules as ``css``.
+
+    Returns
+    -------
+    str
+        The fully qualified component key in the form ``<module_name>.<n>``.
+
+    Raises
+    ------
+    StreamlitAPIException
+        If ``css`` or ``js`` parameters are not strings or None.
+    RuntimeError
+        If the current frame or caller frame cannot be inspected.
+    """
+    # Parameter type guards: only strings or None are supported for js/css
+    for _param_name, _param_value in (("css", css), ("js", js)):
+        if _param_value is not None and not isinstance(_param_value, str):
+            raise StreamlitAPIException(
+                f"{_param_name} parameter must be a string or None. Pass a string path or glob."
+            )
+
+    # Inspect the *call-site* to derive the caller's module and build a fully
+    # qualified component key in the form ``<module_name>.<n>``. This mirrors
+    # the behavior of Components V1 and prevents cross-module name collisions.
+
+    current_frame: FrameType | None = inspect.currentframe()
+    if current_frame is None:
+        raise RuntimeError("Unable to inspect current frame for component declaration.")
+
+    # Go up two frames: current -> component -> user code
+    caller_frame = current_frame.f_back
+    if caller_frame is None:
+        raise RuntimeError(
+            "Unable to determine caller frame for component declaration."
+        )
+
+    user_caller_frame = caller_frame.f_back
+    if user_caller_frame is None:
+        raise RuntimeError(
+            "Unable to determine user caller frame for component declaration."
+        )
+
+    component_key = name
+
+    registry = get_bidi_component_manager()
+    # Check if component was pre-registered from pyproject.toml
+    existing_def = registry.get(component_key)
+
+    if existing_def:
+        # Component exists from pyproject.toml - allow runtime override. Any
+        # provided field (including html) indicates an override operation.
+        if html is not None or css is not None or js is not None:
+            # Create a new definition using exactly the provided values. Fields
+            # omitted by the caller are cleared (set to None) so properties can
+            # be removed via subsequent registrations.
+            # Determine the caller directory to resolve relative paths
```

Comment:
Can you add a sentence here on what the overwrite enables? E.g. does it allow the user to apply runtime changes to an installed component?